src/net.cpp

```diff
@@ -565,41 +565,44 @@ void CNode::copyStats(CNodeStats &stats, const std::vector<bool> &m_asmap)
 }
 #undef X
 
+/**
+ * Try to decode all CNetMessage%s from the first \p nBytes bytes of the \p pch
+ * buffer and send them off for further processing.
+ *
+ * @param[out] complete Whether or not at least one CNetMessage was successfully
+ *                      decoded.
+ *
+ * @returns Whether or not an unhandleable error has occured (e.g. failure to
+ *          deserialize the header or ludicrous message sizes).
+ *
+ * @see ReadMessages(const char *, unsigned int, const CMessageHeader::MessageStartChars&, int64_t)
+ */
 bool CNode::ReceiveMsgBytes(const char *pch, unsigned int nBytes, bool& complete)
 {
-    complete = false;
     int64_t nTimeMicros = GetTimeMicros();
     LOCK(cs_vRecv);
     nLastRecv = nTimeMicros / 1000000;
     nRecvBytes += nBytes;
-    while (nBytes > 0) {
-        // absorb network data
-        int handled = m_deserializer->Read(pch, nBytes);
-        if (handled < 0) return false;
-
-        pch += handled;
-        nBytes -= handled;
 
-        if (m_deserializer->Complete()) {
-            // decompose a transport agnostic CNetMessage from the deserializer
-            CNetMessage msg = m_deserializer->GetMessage(Params().MessageStart(), nTimeMicros);
+    std::tuple<bool, std::vector<CNetMessage>> tup = m_deserializer->ReadMessages(pch, nBytes, Params().MessageStart(), nTimeMicros);
 
-            //store received bytes per message command
-            //to prevent a memory DOS, only allow valid commands
-            mapMsgCmdSize::iterator i = mapRecvBytesPerMsgCmd.find(msg.m_command);
-            if (i == mapRecvBytesPerMsgCmd.end())
-                i = mapRecvBytesPerMsgCmd.find(NET_MESSAGE_COMMAND_OTHER);
-            assert(i != mapRecvBytesPerMsgCmd.end());
-            i->second += msg.m_raw_message_size;
+    std::vector<CNetMessage> msgs = std::get<1>(tup);
```

Comment:
This will copy the vector. Two options:

* Add an `std::move()` around the `std::get` here
* Change the declaration to `std::vector<CNetMessage>& msgs` (or `auto& msgs`).